// scichart/Charting/Visuals/UpdateSuspender.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateSuspender = void 0;
var Dictionary_1 = require("../../Core/Dictionary");
var Guard_1 = require("../../Core/Guard");
/**
 * A class which allows nested suspend/resume operations on an {@link ISuspendable} target
 */
var UpdateSuspender = /** @class */ (function () {
    /**
     * Creates an instance of an {@link UpdateSuspender}
     * @param target The target that we want to suspend. Multiple nested suspend/resumes are permitted
     * @param shouldResumeTarget This flag is passed through to the target on resume
     */
    function UpdateSuspender(target, shouldResumeTarget) {
        if (shouldResumeTarget === void 0) { shouldResumeTarget = true; }
        this.resumeTargetProperty = shouldResumeTarget;
        this.targetProperty = target;
        if (!UpdateSuspender.suspendedInstances.containsKey(this.targetProperty.suspendableId)) {
            UpdateSuspender.suspendedInstances.add(target.suspendableId, { id: target.suspendableId, suspendable: target, suspendCount: 1 });
        }
        else {
            this.increment(target);
        }
    }
    /**
     * Get whether the provided {@link ISuspendable} instance is suspended or not
     * @param target The target {@link ISuspendable}
     */
    UpdateSuspender.getIsSuspended = function (target) {
        return UpdateSuspender.suspendedInstances.containsKey(target.suspendableId);
    };
    /**
     * A helper function to perform multiple operations in a single batch with a single redraw at the end
     * @remarks Implements the .suspendUpdates() and .resume() pattern around a function.
     * Equivalent to calling target.suspendUpdates(), batchOperation() then .resume()
     * @param target The target that we want to suspend and resume around a batch operation
     * @param batchOperation
     */
    UpdateSuspender.using = function (target, batchOperation) {
        Guard_1.Guard.notNull(target, "target");
        var s = target.suspendUpdates();
        try {
            batchOperation();
        }
        finally {
            s.resume();
        }
    };
    Object.defineProperty(UpdateSuspender.prototype, "isSuspended", {
        /**
         * Gets whether the current instance is suspended
         */
        get: function () {
            return UpdateSuspender.getIsSuspended(this.targetProperty);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(UpdateSuspender.prototype, "shouldResumeTarget", {
        /**
         * Gets whether the target should resume once updates are resumed
         */
        get: function () {
            return this.resumeTargetProperty;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Call this to resume drawing on the target {@link ISuspendable}
     */
    UpdateSuspender.prototype.resume = function () {
        this.targetProperty.decrementSuspend();
        if (this.decrement(this.targetProperty) === 0) {
            UpdateSuspender.suspendedInstances.remove(this.targetProperty.suspendableId);
            this.targetProperty.resumeUpdates(this);
        }
    };
    UpdateSuspender.prototype.increment = function (target) {
        UpdateSuspender.suspendedInstances.item(target.suspendableId).suspendCount++;
    };
    UpdateSuspender.prototype.decrement = function (target) {
        var current = UpdateSuspender.suspendedInstances.item(target.suspendableId).suspendCount;
        current--;
        UpdateSuspender.suspendedInstances.item(target.suspendableId).suspendCount = current;
        return current;
    };
    /**
     * USED INTERNALLY: A map of string Id to Suspended instances
     */
    UpdateSuspender.suspendedInstances = new Dictionary_1.Dictionary();
    return UpdateSuspender;
}());
exports.UpdateSuspender = UpdateSuspender;
