// scichart/Core/Thickness.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Thickness = void 0;
/**
 * A class used for margins and padding in SciChart.js
 */
var Guard_1 = require("./Guard");
var Thickness = /** @class */ (function () {
    /**
     * Creates an instance of a Thickness object, with top, right, bottom and left
     * @param top
     * @param right
     * @param bottom
     * @param left
     */
    function Thickness(top, right, bottom, left) {
        this.top = top;
        this.right = right;
        this.bottom = bottom;
        this.left = left;
    }
    /**
     * Creates a Thickness (margin or padding) from string, e.g. "25 50 75 100". Order is top, right, bottom left. Same as in Css
     * @param str
     */
    Thickness.fromString = function (str) {
        Guard_1.Guard.notNull(str, "str");
        var elements = str.split(" ").map(function (s) { return parseFloat(s); });
        if (elements.length !== 4) {
            throw new Error("Expected string in the format '1 2 3 4' where values are top, right bottom and left");
        }
        return new Thickness(elements[0], elements[1], elements[2], elements[3]);
    };
    /**
     * Creates a Thickness (margin or padding) from a single value, e.g. 10, would return a thickness with top, right, bottom left = 10
     * @param value
     */
    Thickness.fromNumber = function (value) {
        return new Thickness(value, value, value, value);
    };
    Thickness.prototype.equals = function (other) {
        return (this.top === other.top &&
            this.right === other.right &&
            this.bottom === other.bottom &&
            this.left === other.left);
    };
    return Thickness;
}());
exports.Thickness = Thickness;
