// scichart/Core/storage/localStorageApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.localStorageApi = void 0;
var checkLocalStorageIsUndefined = function () { return typeof localStorage === "undefined"; };
function getIsLicenseDebug() {
    if (checkLocalStorageIsUndefined())
        return undefined;
    return localStorage.getItem("LICENSE_DEBUG") === "1";
}
function setIsLicenseDebug(value) {
    if (checkLocalStorageIsUndefined())
        return;
    localStorage.setItem("LICENSE_DEBUG", value ? "1" : "0");
}
function clearLicensingDebug() {
    if (checkLocalStorageIsUndefined())
        return;
    localStorage.removeItem("LICENSE_DEBUG");
}
function getLicenseWizardPort() {
    if (checkLocalStorageIsUndefined())
        return undefined;
    var portStr = localStorage.getItem("LICENSE_WIZARD_PORT");
    var portNumber = Number(portStr);
    if (isNaN(portNumber))
        return undefined;
    return portNumber;
}
function getLicenseWizardMaxPort() {
    if (checkLocalStorageIsUndefined())
        return undefined;
    var portStr = localStorage.getItem("LICENSE_WIZARD_MAXPORT");
    var portNumber = Number(portStr);
    if (isNaN(portNumber))
        return undefined;
    return portNumber;
}
exports.localStorageApi = {
    getIsLicenseDebug: getIsLicenseDebug,
    setIsLicenseDebug: setIsLicenseDebug,
    clearLicensingDebug: clearLicensingDebug,
    getLicenseWizardMaxPort: getLicenseWizardMaxPort,
    getLicenseWizardPort: getLicenseWizardPort
};
