// scichart/Charting/Visuals/RenderableSeries/BaseStackedRenderableSeries.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseStackedRenderableSeries = void 0;
var Deleter_1 = require("../../../Core/Deleter");
var guid_1 = require("../../../utils/guid");
var StackedXySeriesInfo_1 = require("../../Model/ChartData/StackedXySeriesInfo");
var IDataSeries_1 = require("../../Model/IDataSeries");
var ResamplingMode_1 = require("../../Numerics/Resamplers/ResamplingMode");
var BaseRenderableSeries_1 = require("./BaseRenderableSeries");
var constants_1 = require("./constants");
var RolloverModifierRenderableSeriesProps_1 = require("./RolloverModifier/RolloverModifierRenderableSeriesProps");
/**
 * Base class for stacked mountain, column series in SciChart's High Performance Real-time
 * {@link https://www.scichart.com/javascript-chart-features | JavaScript Charts}
 * @remarks
 * See derived types {@link StackedMountainRenderableSeries} and {@link StackedColumnRenderableSeries} for
 * details on how to implement stacked column and mountain charts in SciChart
 */
var BaseStackedRenderableSeries = /** @class */ (function (_super) {
    __extends(BaseStackedRenderableSeries, _super);
    /**
     * Creates an instance of a {@link BaseStackedRenderableSeries}
     * @param webAssemblyContext The {@link TSciChart | SciChart WebAssembly Context} containing
     * native methods and access to our WebGL2 WebAssembly Drawing Engine
     */
    function BaseStackedRenderableSeries(webAssemblyContext, options) {
        var _this = _super.call(this, webAssemblyContext, options) || this;
        /** @inheritDoc */
        _this.id = (0, guid_1.generateGuid)();
        /** @inheritDoc */
        _this.isStacked = true;
        _this.rolloverModifierProps = new RolloverModifierRenderableSeriesProps_1.RolloverModifierRenderableSeriesProps(_this);
        // used to track if registered types were used for function properties, so they can be serialized
        _this.typeMap = new Map();
        _this.opacityOriginalValue = 1;
        _this.isStacked = true;
        _this.accumulatedValues = new webAssemblyContext.SCRTDoubleVector();
        _this.accumulatedFinalAnimationValues = new webAssemblyContext.SCRTDoubleVector();
        return _this;
    }
    BaseStackedRenderableSeries.prototype.dataSeriesDataChanged = function () {
        this.notifyPropertyChanged(constants_1.PROPERTY.DATA_SERIES);
    };
    // PUBLIC
    /**
     * @inheritDoc
     */
    BaseStackedRenderableSeries.prototype.delete = function () {
        this.accumulatedValues = (0, Deleter_1.deleteSafe)(this.accumulatedValues);
        this.accumulatedFinalAnimationValues = (0, Deleter_1.deleteSafe)(this.accumulatedFinalAnimationValues);
        // TODO refactor StackedColumnSeries to fix thrown errors
        // super.delete();
        this.dataSeries = (0, Deleter_1.deleteSafe)(this.dataSeries);
        // this.effect = deleteSafe(this.effect);
        this.pointMarker = (0, Deleter_1.deleteSafe)(this.pointMarker);
        this.resamplerHelper = (0, Deleter_1.deleteSafe)(this.resamplerHelper);
        this.pointSeries = (0, Deleter_1.deleteSafe)(this.pointSeries);
        this.dataLabelProviderProperty = (0, Deleter_1.deleteSafe)(this.dataLabelProvider);
    };
    /**
     * Called when the {@link BaseStackedRenderableSeries} is detached from its parent {@link BaseStackedCollection}
     */
    BaseStackedRenderableSeries.prototype.onDetachFromParentCollection = function () {
        // console.log("onDetachFromParentCollection");
        this.parentCollection = undefined;
        this.getParentSurfaceFn = undefined;
        this.notifyParentPropertyChangedFn = undefined;
        this.delete();
    };
    /**
     * @inheritDoc
     */
    BaseStackedRenderableSeries.prototype.notifyPropertyChanged = function (propertyName) {
        if (this.notifyParentPropertyChangedFn) {
            this.notifyParentPropertyChangedFn(propertyName);
        }
    };
    /** @inheritDoc */
    BaseStackedRenderableSeries.prototype.checkIsOutOfDataRange = function (xValue, yValue) {
        var length = this.getDataSeriesValuesCount();
        var isCategoryAxis = this.xAxis.getCurrentCoordinateCalculator().isCategoryCoordinateCalculator;
        var min = isCategoryAxis ? 0 : this.getNativeXValues().get(0);
        var max = isCategoryAxis ? length - 1 : this.getNativeXValues().get(length - 1);
        return xValue < min || xValue > max;
    };
    // NOT SUPPORTED METHODS BEGIN
    /**
     * draw() is not supported for BaseStackedRenderableSeries
     */
    BaseStackedRenderableSeries.prototype.draw = function (renderContext, renderPassData) {
        throw Error("draw() is not supported for BaseStackedRenderableSeries");
    };
    /**
     * getBaseXValues() is not supported for BaseStackedRenderableSeries
     */
    BaseStackedRenderableSeries.prototype.getBaseXValues = function () {
        throw Error("getBaseXValues() is not supported for BaseStackedRenderableSeries");
    };
    /**
     * getXRange() is not supported for BaseStackedRenderableSeries
     */
    BaseStackedRenderableSeries.prototype.getXRange = function () {
        throw Error("getXRange() method is not supported for BaseStackedRenderableSeries");
    };
    /**
     * getYRange() is not supported for BaseStackedRenderableSeries
     */
    BaseStackedRenderableSeries.prototype.getYRange = function (xVisibleRange, isXCategoryAxis) {
        throw Error("getYRange() method is not supported for BaseStackedRenderableSeries");
    };
    /**
     * hasDataSeries() is not supported for BaseStackedRenderableSeries
     */
    BaseStackedRenderableSeries.prototype.hasDataSeries = function () {
        throw Error("hasDataSeries() method is not supported for BaseStackedRenderableSeries");
    };
    /**
     * hasDataSeriesValues() is not supported for BaseStackedRenderableSeries
     */
    BaseStackedRenderableSeries.prototype.hasDataSeriesValues = function () {
        throw Error("hasDataSeriesValues() method is not supported for BaseStackedRenderableSeries");
    };
    /**
     * hasStrokePaletteProvider() is not supported for BaseStackedRenderableSeries
     */
    BaseStackedRenderableSeries.prototype.hasStrokePaletteProvider = function () {
        throw Error("hasStrokePaletteProvider() method is not supported for BaseStackedRenderableSeries");
    };
    /**
     * hasPointMarkerPaletteProvider() is not supported for BaseStackedRenderableSeries
     */
    BaseStackedRenderableSeries.prototype.hasPointMarkerPaletteProvider = function () {
        throw Error("hasFillPaletteProvider() method is not supported for BaseStackedRenderableSeries");
    };
    /**
     * hasFillPaletteProvider() is not supported for BaseStackedRenderableSeries
     */
    BaseStackedRenderableSeries.prototype.hasFillPaletteProvider = function () {
        throw Error("hasFillPaletteProvider() method is not supported for BaseStackedRenderableSeries");
    };
    /**
     * onAttach() is not supported for BaseStackedRenderableSeries
     */
    BaseStackedRenderableSeries.prototype.onAttach = function (scs) {
        throw Error("onAttach() method is not supported for BaseStackedRenderableSeries");
    };
    /**
     * onDetach() is not supported for BaseStackedRenderableSeries
     */
    BaseStackedRenderableSeries.prototype.onDetach = function () {
        throw Error("onDetach() method is not supported for BaseStackedRenderableSeries");
    };
    Object.defineProperty(BaseStackedRenderableSeries.prototype, "animation", {
        /**
         * animation() is not supported for BaseStackedRenderableSeries
         */
        get: function () {
            throw Error("animation() method is not supported for BaseStackedRenderableSeries");
        },
        /**
         * animation() is not supported for BaseStackedRenderableSeries
         */
        set: function (value) {
            throw Error("animation() method is not supported for BaseStackedRenderableSeries");
        },
        enumerable: false,
        configurable: true
    });
    /** @inheritDoc */
    BaseStackedRenderableSeries.prototype.getIndicesRange = function (xRange) {
        throw Error("getIndicesRange() method is not supported for BaseStackedRenderableSeries");
    };
    Object.defineProperty(BaseStackedRenderableSeries.prototype, "resamplingMode", {
        /**
         * resamplingMode property is not supported for BaseStackedRenderableSeries
         */
        get: function () {
            return ResamplingMode_1.EResamplingMode.None;
        },
        set: function (value) {
            throw Error("Setting resamplingMode property is not supported for BaseStackedRenderableSeries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseStackedRenderableSeries.prototype, "resamplingPrecision", {
        /**
         * resamplingPrecision property is not supported for BaseStackedRenderableSeries
         */
        get: function () {
            throw Error("resamplingPrecision property is not supported for BaseStackedRenderableSeries");
        },
        set: function (value) {
            throw Error("resamplingPrecision property is not supported for BaseStackedRenderableSeries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseStackedRenderableSeries.prototype, "parentSurface", {
        // NOT SUPPORTED METHODS END
        /**
         * @inheritDoc
         */
        get: function () {
            return this.getParentSurfaceFn();
        },
        /**
         * set parentSurface property is not supported for BaseStackedRenderableSeries
         */
        set: function (value) {
            throw Error("set parentSurface property is not supported for BaseStackedRenderableSeries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseStackedRenderableSeries.prototype, "xAxis", {
        /** @inheritDoc */
        get: function () {
            return this.parentCollection.xAxis;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseStackedRenderableSeries.prototype, "yAxis", {
        /** @inheritDoc */
        get: function () {
            return this.parentCollection.yAxis;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Runs before the animation starts
     * @protected
     */
    BaseStackedRenderableSeries.prototype.beforeAnimationStart = function () {
        var size = this.accumulatedValues.size();
        this.accumulatedFinalAnimationValues.resize(size, 0);
        for (var i = 0; i < size; i++) {
            this.accumulatedFinalAnimationValues.set(i, this.accumulatedValues.get(i));
        }
        this.opacityOriginalValue = this.opacity;
    };
    /**
     * Runs after the animation is complete
     * @protected
     */
    BaseStackedRenderableSeries.prototype.afterAnimationComplete = function () {
        this.opacityProperty = this.opacityOriginalValue;
    };
    /**
     * Internal method that runs on each animation tick
     * @param progress The current animation progress, a value from 0 to 1
     * @param animationFSM The animation finite state machine
     * @protected
     */
    BaseStackedRenderableSeries.prototype.updateAnimationProperties = function (progress, animationFSM) {
        if (animationFSM.animation.isFadeEffectAnimation) {
            this.opacity = progress * this.opacityOriginalValue;
        }
        else {
            this.opacity = this.opacityOriginalValue;
        }
        animationFSM.animation.calculateAnimationValues(this.webAssemblyContext, this.accumulatedFinalAnimationValues, this.accumulatedValues, progress);
    };
    /** @inheritDoc */
    BaseStackedRenderableSeries.prototype.getSeriesInfo = function (hitTestInfo) {
        return new StackedXySeriesInfo_1.StackedXySeriesInfo(this, hitTestInfo);
    };
    /** @inheritDoc */
    BaseStackedRenderableSeries.prototype.toJSON = function (excludeData) {
        var _a;
        if (excludeData === void 0) { excludeData = false; }
        var options = {
            opacity: this.opacity
        };
        if (((_a = this.dataSeries) === null || _a === void 0 ? void 0 : _a.type) === IDataSeries_1.EDataSeriesType.Xy) {
            // @ts-ignore
            return { type: this.type, options: options, xyData: this.dataSeries.toJSON(excludeData).options };
        }
        else {
            // @ts-ignore
            return { type: this.type, options: options };
        }
    };
    /**
     * toPointSeries method is not supported for BaseStackedRenderableSeries
     * @param resamplingParams
     */
    BaseStackedRenderableSeries.prototype.toPointSeries = function (resamplingParams) {
        throw Error("toPointSeries method is not supported for BaseStackedRenderableSeries");
    };
    /**
     * getCurrentRenderPassData method is not supported for BaseStackedRenderableSeries
     */
    BaseStackedRenderableSeries.prototype.getCurrentRenderPassData = function () {
        throw Error("getCurrentRenderPassData method is not supported for BaseStackedRenderableSeries");
    };
    Object.defineProperty(BaseStackedRenderableSeries.prototype, "xAxisId", {
        /**
         * xAxisId property is not supported for BaseStackedRenderableSeries,
         * instead set on the {@link StackedColumnCollection} or {@link StackedMountainCollection}
         */
        get: function () {
            var _a;
            return (_a = this.parentCollection) === null || _a === void 0 ? void 0 : _a.xAxisId;
        },
        /**
         * xAxisId property is not supported for BaseStackedRenderableSeries,
         * instead set on the {@link StackedColumnCollection} or {@link StackedMountainCollection}
         */
        set: function (value) {
            throw Error("Setting xAxisId property is not supported for BaseStackedRenderableSeries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseStackedRenderableSeries.prototype, "yAxisId", {
        /**
         * yAxisId property is not supported for BaseStackedRenderableSeries,
         * instead set on the {@link StackedColumnCollection} or {@link StackedMountainCollection}
         */
        get: function () {
            throw Error("yAxisId property is not supported for BaseStackedRenderableSeries");
        },
        /**
         * yAxisId property is not supported for BaseStackedRenderableSeries,
         * instead set on the {@link StackedColumnCollection} or {@link StackedMountainCollection}
         */
        set: function (value) {
            throw Error("yAxisId property is not supported for BaseStackedRenderableSeries");
        },
        enumerable: false,
        configurable: true
    });
    BaseStackedRenderableSeries.prototype.onAnimate = function () {
        throw Error("onAnimate is not supported for BaseStackedRenderableSeries");
    };
    Object.defineProperty(BaseStackedRenderableSeries.prototype, "isRunningAnimation", {
        get: function () {
            throw Error("isAnimationRunning is not supported for BaseStackedRenderableSeries");
        },
        enumerable: false,
        configurable: true
    });
    BaseStackedRenderableSeries.prototype.enqueueAnimation = function (animation) {
        throw Error("animationEnqueue is not supported for BaseStackedRenderableSeries");
    };
    BaseStackedRenderableSeries.prototype.runAnimation = function (animation) {
        throw Error("runAnimation is not supported for BaseStackedRenderableSeries");
    };
    return BaseStackedRenderableSeries;
}(BaseRenderableSeries_1.BaseRenderableSeries));
exports.BaseStackedRenderableSeries = BaseStackedRenderableSeries;
