// scichart/Charting/Visuals/RenderableSeries/SplineBandRenderableSeries.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SplineBandRenderableSeries = void 0;
var Deleter_1 = require("../../../Core/Deleter");
var NumberRange_1 = require("../../../Core/NumberRange");
var SeriesType_1 = require("../../../types/SeriesType");
var copyVector_1 = require("../../../utils/copyVector");
var isRealNumber_1 = require("../../../utils/isRealNumber");
var BaseDataSeries_1 = require("../../Model/BaseDataSeries");
var animationHelpers_1 = require("./Animations/animationHelpers");
var BaseBandRenderableSeries_1 = require("./BaseBandRenderableSeries");
var constants_1 = require("./constants");
/**
 * Defines a JavaScript Spline Band-series or High-Low polygon fill chart type in the SciChart's High Performance Real-time
 * {@link https://www.scichart.com/javascript-chart-features | JavaScript Charts}
 * @remarks
 * To add a line series to a {@link SciChartSurface} you need to declare both the {@link SplineBandRenderableSeries | RenderableSeries}
 * and a {@link XyyDataSeries | DataSeries}. Simplified code sample below:
 *
 * ```ts
 * const sciChartSurface: SciChartSurface;
 * const wasmContext: TSciChart;
 * // Create and fill the dataseries
 * const dataSeries = new XyyDataSeries(wasmContext);
 * dataSeries.append(1,2,3);
 * dataSeries.append(2,3,4);
 * // Create the renderableSeries
 * const bandSeries = new SplineBandRenderableSeries(wasmContext);
 * bandSeries.dataSeries = dataSeries;
 * // Set interpolation points
 * bandSeries.interpolationPoints = 10;
 * // append to the SciChartSurface
 * sciChartSurface.renderableSeries.add(bandSeries);
 * ```
 */
var SplineBandRenderableSeries = /** @class */ (function (_super) {
    __extends(SplineBandRenderableSeries, _super);
    /**
     * Creates an instance of the {@link SplineBandRenderableSeries}
     * @param webAssemblyContext The {@link TSciChart | SciChart WebAssembly Context} containing
     * native methods and access to our WebGL2 WebAssembly Drawing Engine
     * @param options optional parameters of type {@link ISplineBandRenderableSeriesOptions} applied when constructing the series type
     */
    function SplineBandRenderableSeries(webAssemblyContext, options) {
        var _this = this;
        var _a, _b, _c;
        _this = _super.call(this, webAssemblyContext, options) || this;
        _this.type = SeriesType_1.ESeriesType.SplineBandSeries;
        _this.warnOnSplineFailure = true;
        _this.isSplineProperty = true;
        _this.recalculateSpline = true;
        _this.interpolationPointsProperty = 10;
        _this.xSplineValues = new webAssemblyContext.SCRTDoubleVector();
        _this.ySplineValues = new webAssemblyContext.SCRTDoubleVector();
        _this.y1SplineValues = new webAssemblyContext.SCRTDoubleVector();
        _this.interpolationPointsProperty = (_a = options === null || options === void 0 ? void 0 : options.interpolationPoints) !== null && _a !== void 0 ? _a : _this.interpolationPointsProperty;
        // Must be called here for the series type to be available
        if ((_b = _this.paletteProvider) === null || _b === void 0 ? void 0 : _b.onAttached) {
            (_c = _this.paletteProvider) === null || _c === void 0 ? void 0 : _c.onAttached(_this);
        }
        return _this;
    }
    Object.defineProperty(SplineBandRenderableSeries.prototype, "isSpline", {
        get: function () {
            return this.isSplineProperty;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SplineBandRenderableSeries.prototype, "interpolationPoints", {
        /**
         * Gets or sets the interpolationPoints being used for the Spline
         */
        get: function () {
            return this.interpolationPointsProperty;
        },
        /**
         * Gets or sets the interpolationPoints being used for the Spline
         */
        set: function (value) {
            this.interpolationPointsProperty = value;
            this.notifyPropertyChanged(constants_1.PROPERTY.INTERPOLATION_POINTS);
        },
        enumerable: false,
        configurable: true
    });
    /** @inheritDoc */
    SplineBandRenderableSeries.prototype.delete = function () {
        this.xSplineValues = (0, Deleter_1.deleteSafe)(this.xSplineValues);
        this.ySplineValues = (0, Deleter_1.deleteSafe)(this.ySplineValues);
        this.y1SplineValues = (0, Deleter_1.deleteSafe)(this.y1SplineValues);
        _super.prototype.delete.call(this);
    };
    /**
     * Updates spline values
     */
    SplineBandRenderableSeries.prototype.updateSplineValues = function () {
        if (!(this.dataSeries && this.recalculateSpline && this.parentSurface)) {
            return;
        }
        this.isSplineProperty = true;
        var dataSeries = this.dataSeries;
        if (dataSeries.fifoCapacity) {
            throw new Error("Sorry, fifo is not currently supported for spline series");
        }
        var xValues = this.xAxis.isCategoryAxis ? dataSeries.getNativeIndexes() : dataSeries.getNativeXValues();
        var yValues = dataSeries.getNativeYValues();
        var y1Values = dataSeries.getNativeY1Values();
        var initialSize = xValues.size();
        this.webAssemblyContext.SCRTSplineHelperCubicSpline(xValues, yValues, this.xSplineValues, this.ySplineValues, initialSize, this.interpolationPoints, dataSeries.dataDistributionCalculator.containsNaN);
        if (isNaN(this.ySplineValues.get(0))) {
            this.onSplineFailure();
        }
        else {
            this.webAssemblyContext.SCRTSplineHelperCubicSpline(xValues, y1Values, this.xSplineValues, this.y1SplineValues, initialSize, this.interpolationPoints, dataSeries.dataDistributionCalculator.containsNaN);
        }
        this.recalculateSpline = false;
    };
    /** @inheritDoc */
    SplineBandRenderableSeries.prototype.onSplineFailure = function () {
        if (this.warnOnSplineFailure) {
            console.warn("Could not calculate spline values.  X data may contain duplicates.  Falling back to original values.\nTo disable this warning set warnOnSplineFailure = false.  To change this behaviour, override onSplineFailure");
        }
        var xValues = this.xAxis.isCategoryAxis
            ? this.dataSeries.getNativeIndexes()
            : this.dataSeries.getNativeXValues();
        (0, copyVector_1.copyDoubleVector)(xValues, this.xSplineValues, this.webAssemblyContext);
        (0, copyVector_1.copyDoubleVector)(this.dataSeries.getNativeYValues(), this.ySplineValues, this.webAssemblyContext);
        (0, copyVector_1.copyDoubleVector)(this.dataSeries.getNativeY1Values(), this.y1SplineValues, this.webAssemblyContext);
        // This tells the drawing providers not to treat the data as interpolated
        this.isSplineProperty = false;
    };
    /** @inheritDoc */
    SplineBandRenderableSeries.prototype.getYRange = function (xVisibleRange, isXCategoryAxis) {
        var _a;
        if (isXCategoryAxis === void 0) { isXCategoryAxis = false; }
        if (this.isRunningAnimation) {
            return this.dataSeries.getWindowedYRange(xVisibleRange, true, isXCategoryAxis);
        }
        this.updateSplineValues();
        var xValues = this.xSplineValues;
        var yValues = this.ySplineValues;
        var y1Values = this.y1SplineValues;
        // TODO: getPositiveRange
        // if one point
        if (xValues.size() === 1) {
            var min = Math.min(yValues.get(0), y1Values.get(0));
            var max = Math.max(yValues.get(0), y1Values.get(0));
            return new NumberRange_1.NumberRange(min, max);
        }
        var indicesRange = isXCategoryAxis
            ? xVisibleRange
            : (0, BaseDataSeries_1.getIndicesRange)(this.webAssemblyContext, xValues, xVisibleRange, (_a = this.dataSeries) === null || _a === void 0 ? void 0 : _a.dataDistributionCalculator.isSortedAscending);
        var iMin = Math.max(Math.floor(indicesRange.min), 0);
        var iMax = Math.min(Math.ceil(indicesRange.max), xValues.size() - 1);
        if (iMax < iMin) {
            return undefined;
        }
        var minMax;
        var minMaxy1;
        try {
            minMax = this.webAssemblyContext.NumberUtil.MinMaxWithIndex(yValues, iMin, iMax - iMin + 1);
            if (!(0, isRealNumber_1.isRealNumber)(minMax.minD) || !(0, isRealNumber_1.isRealNumber)(minMax.maxD)) {
                return undefined;
            }
            minMaxy1 = this.webAssemblyContext.NumberUtil.MinMaxWithIndex(y1Values, iMin, iMax - iMin + 1);
            if (!(0, isRealNumber_1.isRealNumber)(minMaxy1.minD) || !(0, isRealNumber_1.isRealNumber)(minMaxy1.maxD)) {
                return undefined;
            }
            return new NumberRange_1.NumberRange(Math.min(minMax.minD, minMaxy1.minD), Math.max(minMax.maxD, minMaxy1.maxD));
        }
        finally {
            (0, Deleter_1.deleteSafe)(minMax);
            (0, Deleter_1.deleteSafe)(minMaxy1);
        }
    };
    /** @inheritDoc */
    SplineBandRenderableSeries.prototype.toJSON = function (excludeData) {
        if (excludeData === void 0) { excludeData = false; }
        var json = _super.prototype.toJSON.call(this, excludeData);
        var options = {
            interpolationPoints: this.interpolationPoints
        };
        Object.assign(json.options, options);
        return json;
    };
    /** @inheritDoc */
    SplineBandRenderableSeries.prototype.setAnimationVectors = function (animation) {
        var ds = this.dataSeries;
        if (!ds.xInitialAnimationValues) {
            ds.createAnimationVectors();
        }
        this.updateSplineValues();
        animationHelpers_1.animationHelpers.setSplineBandAnimationVectors(this.webAssemblyContext, animation, this.dataSeries, animation.dataSeries, this.xSplineValues, this.ySplineValues, this.y1SplineValues, this.xAxis.isCategoryAxis, this.interpolationPoints);
    };
    /** @inheritDoc */
    SplineBandRenderableSeries.prototype.updateAnimationProperties = function (progress, animationFSM) {
        var animation = animationFSM.animation;
        animation.updateSeriesProperties(this, animationFSM.initialStyles, animationFSM.animationProgress);
        var dataSeries = this.dataSeries;
        if (dataSeries) {
            animationHelpers_1.animationHelpers.updateSplineBandAnimationProperties(this.webAssemblyContext, animation, progress, dataSeries, this.xSplineValues, this.ySplineValues, this.y1SplineValues);
        }
        if (this.invalidateParentCallback) {
            this.invalidateParentCallback();
        }
    };
    /** @inheritDoc */
    SplineBandRenderableSeries.prototype.dataSeriesDataChanged = function () {
        this.recalculateSpline = true;
        _super.prototype.dataSeriesDataChanged.call(this);
    };
    return SplineBandRenderableSeries;
}(BaseBandRenderableSeries_1.BaseBandRenderableSeries));
exports.SplineBandRenderableSeries = SplineBandRenderableSeries;
